### DISCUSSION TITLE
Issue with ComputeFiniteStrain - seems unable to solve problem with zero rotation

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#27292] [Q&A Modules: Solid mechanics] [answered]
Hi, I am solving a simple problem which involves stretching of a membrane in one direction. I get the following error with `ADComputeFiniteStrain`:

```
Cannot take square root of a number less than or equal to zero in the calculation of C1 for the Rashid approximation for the rotation tensor. This zero or negative number may occur when elements become heavily distorted.
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
```

However, if I switch to ComputeIncrementalSmallStrain it works fine. I initially thought the issue was with the stretch ratio of 1.1 (i.e 1.1 times it's original size) being to small to be considered Finite Strain, but then when I increased the ratio to 1.8, I got the same error. I am using an incredibly fine mesh, and since the poisson's ratio is close to 0.5, I also applied a volumetric locking correction. 

I am not sure how to solve this issue. There is no rotation in my problem so the rotation tensor *should* be zero.

**Comment by GiudGiud** [comment]
@gambka @jiangwen84

**Comment by dschwen** [comment]
> ComputeFiniteStrain - seems unable to solve problem with zero rotation

If "no rotation" really was the issue then we wouldn't even be able to compute the initial residual or a body at rest with no loading. I'm pretty sure that's _not_ the issue here. It would help if you added a minimal non-working example here.

**Reply by richmondodufisan** [reply]
I see. Here's the input file with the issue:

https://github.com/richmondodufisan/purple/blob/main/Cornea_Linear_Elastic/Cornea_Stretch.i

**Reply by richmondodufisan** [reply]
Hi @dschwen , were you able to take a look? I haven't been able to figure out the reason for the error

**Comment by gambka** [comment]
What order elements are you using? The `volumetric_locking_correction` is not required for higher (second) order elements. What happens if you set the `decomposition_method = EigenSolution` instead of the default `TaylorExpansion`?

**Reply by richmondodufisan** [reply]
I was using using first order elements. When I switched the `decomposition_method`, it didn't converge, the non-linear residual stayed at the same order of magnitude. I tried switching to 2nd order elements (and I took out the `volumetric_locking_correction`) but it still did the same thing. Here is the input file:

https://github.com/richmondodufisan/purple/blob/main/Cornea_Linear_Elastic/Cornea_Stretch.i

**Reply by richmondodufisan** [reply]
Hi @gambka , were you able to take a look? I haven't been able to figure out the issue yet

**Comment by hugary1995** [comment]
A few suggestions:
1. Ramp up displacement bc with time -- allow the solve to fail and time step to be cut.
2. Set preset=false in the bcs.
3. Turn off line search.

**Reply by richmondodufisan** [reply]
Just came back to this, and setting `preset = false` worked! thanks! I checked the documentation and it says that determines `Whether or not to preset the BC (apply the value before the solve begins)`. I'm not fully sure what that means, could you explain further? For a steady state solve, don't the BCs have to be prescribed before solving the Kd=F system? Since the system isn't solved by direct inversion and is instead solved iteratively, does this mean on the first iteration no BCs are defined? Would it be good practice to always set `preset=false` (i.e is one more accurate than the other?)?

**Reply by GiudGiud** [reply]
It means whether to just set the degrees of freedom (preset = true, default) for the boundary displacements directly and take them off the system solve
or whether to solve for them

the difference is that setting them directly can destroy the boundary elements, making their volume negative for example

**Reply by richmondodufisan** [reply]
Hi @hugary1995 @GiudGiud 

I tried to run a slightly modified version of this problem, i.e I didn't restrain the top and bottom of the plate (they were previously restrained), and pulled it apart horizontally. Basically a more realistic uniaxial tension test. I couldn't get a convergent solution with ComputeFiniteStrain. 

I found this old thread https://github.com/idaholab/moose/discussions/20285 where @cbolisetti mentioned that MOOSE cannot model nonlinearities yet. Could this be the issue? Since I expect geometric nonlinearities due to the large deformation (material is incompressible and I'm applying stretch values of up to 1.8, so I expect it to deform considerably)

**Reply by GiudGiud** [reply]
> I found this old thread https://github.com/idaholab/moose/discussions/20285 where @cbolisetti mentioned that MOOSE cannot model nonlinearities yet

this was likely in a very specific context. We use a nonlinear solve for most things.

>  I didn't restrain the top and bottom of the plate (they were previously restrained), 

This is likely the problem. Please make sure there is no parasitic translation or rotation mode. If they are in the nullspace, then any solution with any amount of these modes is still a solution. Some solvers dont like that

**Reply by hugary1995** [reply]
First off, the solid mechanics module definitely can handle geometric nonlinearities, that's in a sense all we do... @cbolisetti was meant to say certain types of elements (shell, beam, etc) do not have such support yet, which doesn't seem like what you are using anyways.

Now back to your question: Theoretically speaking, if your material model satisfies polyconvexity, the material model Jacobian is exact, and the boundary condition increment is small enough within the current step, it's guaranteed to converge. Several things could hinder convergence in practice:
1. You are using a custom material model -- that we can't really help, as it's the user's responsibility of using a well formed material model.
2. The custom material model has an inexact Jacobian -- use AD or hand calculate the Jacobian more carefully.
3. The solid mechanics module is giving you inexact Jacobian outside the material model, which is possible with the old system -- using AD should fix that as well.
4. BC increment not small enough -- let the solve fail, and MOOSE should reattempt the solve with a small time step size.
5. Turning off line search might help with plasticity, nearly incompressible material, etc.
6. Preset BC might make things worse, as we discussed before.

**Reply by hugary1995** [reply]
> > I didn't restrain the top and bottom of the plate (they were previously restrained),
> 
> This is likely the problem. Please make sure there is no parasitic translation or rotation mode. If they are in the nullspace, then any solution with any amount of these modes is still a solution. Some solvers dont like that

That is a very good point as well.

**Reply by richmondodufisan** [reply]
This is a rough sketch of what I'm expecting:

![tension](https://github.com/user-attachments/assets/ae64fb32-1dd6-4ecb-831b-c34cd5da06bf)

 if I restrain the top and bottom then I won't be able to get this deformed shape. In the input file https://github.com/richmondodufisan/purple/blob/main/Cornea_Linear_Elastic/Cornea_Stretch_Incremental_Load.i 

I set preset = false on all BCs
I only used MOOSE provided kernels and material models (ADComputeFiniteStrain, ADComputeFiniteStrainElasticStress, etc)
I ramped up the displacement BC up to 1/100 of the desired value
The material is nearly incompressible, so I set line search = none and also set volumetric_locking_correction = true
I am using a very fine mesh with second order elements (first order also did not converge)

What else can I do to get a convergent solution?

**Reply by GiudGiud** [reply]
could you restrain a node on the left center, a node on the center of the right surface, and the very center node
I think this would not change the solution at all? and it would kill a number of translational and rotational modes. You'd still miss the rotation around the center axis

**Reply by richmondodufisan** [reply]
Instead of pulling on both sides, to model this I completely restrained the left side in x and y, and I'm pulling on the right side (applied displacement on the right side only). So that means restraining the center node also would not work.

I also toggled restraining the right side in y (vertical) but it made no difference.

**Reply by GiudGiud** [reply]
> I completely restrained the left side in x and y,

so is the thinning of the cylinder possible in those conditions?

**Reply by richmondodufisan** [reply]
yes, only the boundary is restrained, the rest of the material is free to move wherever like this:

![4978864529660292242](https://github.com/user-attachments/assets/32f7bddd-fe0c-4f05-b5ec-f547efe24dba)

**Reply by GiudGiud** [reply]
ok seems fair to me.
Is it blocked in x,y and z though at the center?

**Reply by richmondodufisan** [reply]
Do you think it should be? I would expect the y displacement to be constant at the center, but with this configuration the x displacement should gradually decrease from the prescribed value to zero from right to left.

Also, how would I implement a BC for a single node/point? DirichletBC requires a boundary

**Reply by GiudGiud** [reply]
>  but with this configuration the x displacement should gradually decrease from the prescribed value to zero from left to right.

umm shouldnt it be just on the very left that the boundary is set?
I mean x,y and z on the surface

NodalBCs do single node values

**Reply by cbolisetti** [reply]
If the material is almost incompressible, do you see the 'dog bone' type deformed shape? I thought you need a realistic poisson's ratio to see that.

**Reply by richmondodufisan** [reply]
yes the fixed boundary is set on the very left and fixed displacement is set on the very right. but the solution should be the x-displacement dropping from the fixed value to zero from right to left (while the y displacement is maximum at the top and bottom).

Also it's 2D so there's only x and y components.

There is no NodalBC object here https://mooseframework.inl.gov/syntax/BCs/, just DirichletBC

**Reply by richmondodufisan** [reply]
@cbolisetti good point! yeah I think that is the issue. I just changed the poisson's ratio to a lower value and it converged fine. I find it interesting though that if I set it to nearly incompressible, then I need to also fix the top and bottom boundaries for it to converge. But yeah I think that was the source of the issue, thanks! Rubber might not be the best approximation for my material then...

**Reply by GiudGiud** [reply]
> There is no NodalBC object here https://mooseframework.inl.gov/syntax/BCs/, just DirichletBC

I think a dirichlet BC on a nodeset of a single node is a nodal bc.
Otherwise a nodal kernel will do as well.

**Reply by richmondodufisan** [reply]
> If the material is almost incompressible, do you see the 'dog bone' type deformed shape? I thought you need a realistic poisson's ratio to see that.

Actually @cbolisetti could you expand on why you think so? I was just thinking now that if the material is nearly incompressible, the stretch product lambda_1 * lambda_2 = 1. And if so, then lambda_2 < 1 in the x2 direction to compensate for the applied stretch. Hence fixing the top and bottom surfaces shouldn’t be right even though that’s the only way I’m getting it to converge with poussins ratio of 0.4999

**Reply by hugary1995** [reply]
I suggest you try the new Lagrangian mechanics system with nu = 0.4999, free top/bottom boundaries, and volumetric locking correction. Take a look at the regression tests `modules/solid_mechanics/test/tests/lagrangian` to see how to set it up. Try the Jaumann and Green-Naghdi objective rates. I don't see any reason for it to fail, but in case it does, post the input file here and I can take a look.

As a general suggestion, hyperelasticity might be more suitable for this type of incompressible materials.

**Reply by richmondodufisan** [reply]
Okay thank you, I’ll give it a shot and let you know. And yes I plan to eventually implement a hyperelastic model (when I figure out how to do that!), I just wanted to check that the dispersion curve I’m getting for a pre-stretched linear elastic material lines up with the analytical solution first.

**Reply by hugary1995** [reply]
Don't you need to run some dynamic simulations to extract the dispersion curve, or are we talking about different dispersion curves?

**Reply by richmondodufisan** [reply]
Yes, but that’s in a separate input file that uses the results of this simple stretch problem. Also I’m running them in the frequency domain there. But I gotta get this stretch to work first and then I’ll take it from there

**Reply by hugary1995** [reply]
Okay, that makes sense. Keep me posted.

**Reply by richmondodufisan** [reply]
@hugary1995 

Hi, I implemented with the new system. I couldn't get it to converge without restraining the top and bottom boundaries. 

-The "base_name" functionality didn't work, I had to hardcode the name of the elasticity tensor in the input file
-I set large_kinematics = true globally
-I also set stabilize_strain = true globally, for volumetric locking correction
-It didn't converge, so I tried switching to second order elements but it also did not converge

In all simulations, the residual would decrease and then remain constant for several iterations until the "solve did not converge" error shows up. Then I tried setting the displacement smaller, and the lower the applied displacement, the lower (and closer to convergence) the residual would get. When the displacement became small enough (1%, i.e stretch_ratio = 1.001), it converged as it was essentially a small strain problem.

Does this mean that I cannot get a convergent result unless I use a hyperelastic model? I know that for large stretches, using a linear elastic model is unphysical but I didn't expect it to not converge at all. Also, looking at the dimensions of modules/solid_mechanics/test/tests/lagrangian/cartesian/total/patch/large_patch.i, the applied displacement seems greater than 1% of the mesh dimensions.

Here is the relevant input file: https://github.com/richmondodufisan/purple/blob/main/Cornea_Linear_Elastic/Cornea_Stretch.i

*********************************************************************************************************************************

If this is all true I want to try implementing a hyperelastic model. I read through the documentation but want to ask about my approach before trying. The steps I plan to take are:

-Implement a custom strain energy function, W(F), with a custom material
-Implement a material that inherits from ComputeLagrangianStressPK1, and define the stress as the derivative of W wrt F

-To calculate the derivative, I want to follow the steps in https://mooseframework.inl.gov/source/materials/DerivativeMaterialInterface.html#:~:text=The%20derivative%20of%20a%20material,derivatives%20of%20other%20material%20properties.

In this case, I will "get" the custom strain energy material I created, and differentiate wrt "deformation_gradient" (the declared name of F as seen in ComputeLagrangianStrainBase.C)

Am I missing any steps?

**Reply by richmondodufisan** [reply]
@hugary1995 did you get a chance to give this a look? your feedback will be much appreciated

**Reply by richmondodufisan** [reply]
Update:

Implemented the hyperelastic (neo Hookean) model as described, but again, cannot get convergence past 1% strain for this incompressible formulation.

Input file: test_3.i
https://github.com/richmondodufisan/purple/tree/main/Cornea_Hyperelastic_NeoHookean

![disp_x](https://github.com/user-attachments/assets/128f6bfb-4b34-4832-8b4a-f916faf229c5)


![disp_y_newtol](https://github.com/user-attachments/assets/dc9c3088-9ad2-461e-bca9-733522edf1a3)

**Reply by richmondodufisan** [reply, marked as answer]
Hi, I found the source of the problem. To get it to converge to the right solution I:

-Switched to manual numerical differentiation, was able to get at least one converged timestep. Apparently, the `getMaterialPropertyDerivative` cannot do tensor differentiation.

-Switched to a penalty method. The `PC failed due to FACTOR_NUMERIC_ZEROPIVOT` was coming from the extra field (pressure), and there was no boundary condition to set to it. After implementing that that went away.

-Switched to PK2 formulation (probably didnt do as much to help)

-Set low floating point values to zero, they were coming up in the deformation gradient and seemed to throw the solution in the opposite direction? (see the last result above). I'm not sure how but I remember this being the last thing I did before it worked.


![correct_x](https://github.com/user-attachments/assets/c7528ec2-f1f5-4f2a-b45c-82bb7dc6f0ea)


![correct_y](https://github.com/user-attachments/assets/23426e20-ae89-4521-9046-3486597cf61d)


All this was because I couldn't get a similar problem to converge for MOOSE's linear elastic material for any deformation more than 1%. Maybe using a nearly incompressible poissons ratio on a linear elastic material really is that stiff even with stabilize_strain?

**Reply by GiudGiud** [reply]
@richmondodufisan for item 1 and 4 did you modify the code to do this?
for item 2 and 3, this was done in the input file?
Asking because @Spudman105 is hitting similar difficulties in https://github.com/idaholab/moose/discussions/30483
Can you share your modifications?

**Reply by richmondodufisan** [reply]
Yes sure. In hindsight I'm not sure how much of an impact setting low floating point numbers to zero had, since I implemented all those four changes at the same time.

From my POV the biggest change I made was to switch to the new system and implement my own kernel & material. I have since gone further to manually code the tangent operator, and so the numerical differentiation I did with a for-loop + finite difference is no longer relevant.

Here are my input files for all three cases (nearly incompressible, incompressible, and compressible). The custom kernels are also in the same repository.

https://github.com/richmondodufisan/purple/tree/main/Cornea_Hyperelastic_NeoHookean/stress_strain_tests

However, looking at their input file I am not sure why they would get that error, because through several trial/error attempts I figured out that the source that error (for me) was from the nearly incompressible poisson ratio, which ComputeFiniteStrain (old system) was unable to handle, even with volumetric_locking_correction set to true. When I reduced the poisson ratio it solved easily. 

They seem to be dealing with a compressible problem, and I think their error is likely coming from something else (maybe the BCs are too extreme or something, I'd suggest they try to replace the applied displacements with very, very small values so it can solve/converge and they can look at the paraview output for clues).
